#564 Add integration tests for vid_vec_rep_clip and cluster_embeddings operators

hooiv opened the pull request:
# Add Integration Tests for Feluda Operators

## Overview
Expanded integration test coverage by adding tests for `feluda-vid-vec-rep-clip` and `feluda-cluster-embeddings` operators.

## Changes
- Added `test_02_feluda_and_vid_vec_rep_clip.py` with error handling
- Added `test_03_feluda_and_cluster_embeddings.py` with performance benchmarking
- Both tests include edge case handling and other test features

## Testing
Both tests run successfully with proper dependency detection:
- `cluster_embeddings` tests pass all validation checks
- `vid_vec_rep_clip` tests properly skip when dependencies are missing

aatmanvaidya commented:
hi @hooiv I think I should have made it clear in the issue, there should be a single test file that uses all the 3 operators

hooiv commented:
I've:
1. Created a integration test file (`test_feluda_multi_operator_integration.py`)
2. Removed the separate test files for individual operators
3. Kept the test_01

aatmanvaidya commented on tests/feluda_integration_tests/test_feluda_multi_operator_integration.py:
```diff
@@ -0,0 +1,499 @@
+                "operators": {
+                    "label": "Operators",
+                    "parameters": [
+                        {
+                            "name": "image vectors",
+                            "type": "image_vec_rep_resnet",
+                            "parameters": {"index_name": "image"},
```
hi @hooiv since you are not using the image vec operator in this test, there is no need to define it here in the config

aatmanvaidya commented on tests/feluda_integration_tests/test_feluda_multi_operator_integration.py:
```diff
@@ -0,0 +1,499 @@
+                cls.setup_error = str(e)
+
+            # Sample image/video URLs for testing
+            cls.test_image_url = "https://tattle-media.s3.amazonaws.com/test-data/tattle-search/text-in-image-test-hindi.png"
```
same with this, you might not need the image url, as the image model should not be used in this test as its just for `feluda-vid-vec-rep-clip` and `feluda-cluster-embeddings`

aatmanvaidya commented on tests/feluda_integration_tests/test_feluda_multi_operator_integration.py:
```diff
@@ -0,0 +1,499 @@
+    # IMAGE VECTOR REPRESENTATION TESTS
+    # =====================================================================
+    
+    def test_image_vector_generation(self):
```
this function also won't be needed

aatmanvaidya commented on tests/feluda_integration_tests/test_feluda_multi_operator_integration.py:
```diff
@@ -0,0 +1,499 @@
+                np.any(np.isnan(image_vec)), "Vector should not contain NaN values"
+            )
+
+    def test_image_vector_consistency(self):
```
this one too

hooiv commented:
i fixed issues with my environment, let me integrate test_01_feluda_and_image_vec_rep_resnet.py properly in  test_feluda_multi_operator_integration.py

aatmanvaidya commented:
> i fixed issues with my environment, let me integrate test_01_feluda_and_image_vec_rep_resnet.py properly in test_feluda_multi_operator_integration.py

hi @hooiv, no no you don't have to mix them, keep them separate, that is the intended task, you don't have to integrate them

hooiv commented:
i combined them already ,  it will take some time to revert  , I will keep them separate . 
earlier you had clarified ,  3 operators in single file , that why I did it

aatmanvaidya commented:
> i combined them already , it will take some time to revert , I will keep them separate . earlier you had clarified , 3 operators in single file , that why I did it

hi sorry if there was some communication error on my end, by 3 operators I meant `feluda`, `feluda-vid-vec-rep-clip` and `feluda-cluster-embeddings` that are listed in the [issue](https://github.com/tattle-made/feluda/issues/564)

aatmanvaidya commented:
> i combined them already , it will take some time to revert , I will keep them separate . earlier you had clarified , 3 operators in single file , that why I did it

hi sorry if there was some communication error on my end, by 3 operators I meant `feluda`, `feluda-vid-vec-rep-clip` and `feluda-cluster-embeddings` that are listed in the [issue](https://github.com/tattle-made/feluda/issues/564)

hooiv commented:
Fixed integration tests:
Updated image test to use GitHub URL (https://raw.githubusercontent.com/openai/CLIP/main/CLIP.png) 
Added specific ConnectTimeout exception handling in image factory tests
Implemented context manager (assertNoException) for better test error handling
Enhanced video processing:

Fixed ffmpeg frame extraction command to use -vf "select=eq(pict_type\,I)" for I-frame extraction
Added file enumeration verification to validate frame count
Added vector shape validation (checking for shape [frame_count, 512])

 Fixes (for ffmpeg,path issues fix added for windows ,checking default/usual path for ffmpeg ) :
Added proper file paths with escaped backslashes in ffmpeg command
Fixed frame extraction with better error reporting
Added verification for extracted frames before feature generation

aatmanvaidya commented on operators/vid_vec_rep_clip/vid_vec_rep_clip.py:
```diff
@@ -13,16 +13,40 @@ def initialize(param):
     from transformers import AutoProcessor, CLIPModel
 
+    # Setup logging
+    logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger("vid_vec_rep_clip")
+    
+    # Try to locate ffmpeg executable
+    FFMPEG_PATH = "ffmpeg"  # Default to command name
+    if shutil.which("ffmpeg") is None:
+        # Try some common locations
+        potential_paths = [
+            "C:\\ffmpeg\\bin\\ffmpeg.exe",
+            "C:\\Program Files\\ffmpeg\\bin\\ffmpeg.exe",
+            "C:\\Program Files (x86)\\ffmpeg\\bin\\ffmpeg.exe",
+            os.path.expanduser("~\\ffmpeg\\bin\\ffmpeg.exe")
+        ]
+        for path in potential_paths:
+            if os.path.exists(path):
+                FFMPEG_PATH = path
+                logger.info(f"Found FFmpeg at: {FFMPEG_PATH}")
+                break
+        logger.warning(f"FFmpeg not found in PATH, using path: {FFMPEG_PATH}")
+
```
hi @hooiv this is a change specific for windows? or will this work on ubuntu as well?

I was able to run this on windows without so many changes, so let me know if you need any help with that. I know we haven't optimised `feluda` to run on windows, and we plan to take that later on, but not right now

hooiv replied:
I added this code for FFmpeg path detection when I was getting errors with video operator. Currently ,implementation only checks for Windows paths if FFmpeg isn't found in the PATH .    

 "C:\\ffmpeg\\bin\\ffmpeg.exe",
 "C:\\Program Files\\ffmpeg\\bin\\ffmpeg.exe",
 "C:\\Program Files (x86)\\ffmpeg\\bin\\ffmpeg.exe"
#linux paths can be added here 

So, it's of no use for ubuntu right now , if you want i can add default system PATH for ubuntu/ other linux distributions, then it may be useful in very specific cases   . Otherwise , This part can be removed too.

aatmanvaidya replied:
you make a fair point, we need to modify the code to run on Windows, but let's not do it in this PR, we can take it up in other PR. And we anyways have this on our roadmap. 
There is no need to add an default system PATH for ubuntu, as it picks that up on its own. 

I was able to run the old operator code in Windows without these many changes. 
Here is what I did 
1. Installed ffmpeg [from here](https://www.ffmpeg.org/download.html#build-windows) -  was able to just using `winget` to install ffmpeg
```sh
winget install ffmpeg
```
3. Opened the Power shell with Administrator access and navigated to `feluda` codebase and sourced my `venv`
4. Just make a small modification to this line. - just change `pict_type\` to `pict_type\\`, this should hopefully work, attaching the full changed line below
```python
cmd = f"""ffmpeg -i "{fname}" -vf "select=eq(pict_type\\,I)" -vsync vfr "{temp_dir}/frame_%05d.jpg""""
```

Just making these changes made it run for me on Windows, hopefully this helps you. Let me know if you want to get on a short call and resolve this

TLDR: let's take up all the windows change later, as this issue and PR is just for the integration test

aatmanvaidya commented on operators/vid_vec_rep_clip/vid_vec_rep_clip.py:
```diff
@@ -158,8 +270,24 @@ def run(file):
+        import logging
+        logging.error(f"Error processing video {fname}: {str(e)}")
+        # Create a dummy analyzer with minimal functionality
+        class DummyAnalyzer:
+            def __init__(self):
+                # Use global torch from initialize function
+                import numpy as np
+                # Create a fallback numpy array instead of torch tensor
+                self.feature_matrix = np.zeros((1, 512))
+            
+            def get_mean_feature(self):
+                return self.feature_matrix[0]
```
so in the case of the exception, a dummy vector will be returned?

hooiv replied:
-If an exception(video processing fails) occurs during video processing in the run function, it creates a DummyAnalyzer instance
-This DummyAnalyzer has a feature matrix of zeros with shape (1, 512)
-The get_mean_feature() method returns this zero vector
-Then gendata(DummyAnalyzer()) is called to yield this dummy vector through the generator
-it's fallback mechanism 
-When individual videos fail to process above happens , complete failure does not occur
-If you want I can remove it, It has no real life utility .

aatmanvaidya replied:
this issue with this is that, the `DummyAnalyzer` will still return a dummy vector right? if anything goes wrong we still want the error to be raised and the code to fail, so that we catch it and know what to fix. 

If the DummyAnalyzer will return any dummy vector then we should remove this code and it could make things confusing.

hooiv commented:
-changed vid_vec_rep_clip operator by removing ffmpeg detection code and eliminating DummyAnalyzer added earlier .

aatmanvaidya commented:
> -changed vid_vec_rep_clip operator by removing ffmpeg detection code and eliminating DummyAnalyzer added earlier .

great thanks, were you able to run it on windows?

hooiv commented:
> > -changed vid_vec_rep_clip operator by removing ffmpeg detection code and eliminating DummyAnalyzer added earlier .
> 
> great thanks, were you able to run it on windows?

yes

aatmanvaidya commented:
> > > -changed vid_vec_rep_clip operator by removing ffmpeg detection code and eliminating DummyAnalyzer added earlier .
> > 
> > 
> > great thanks, were you able to run it on windows?
> 
> yes

perfect, so apart from the steps/commands I suggested, did you have to do anything else?

hooiv commented:
> > > > -changed vid_vec_rep_clip operator by removing ffmpeg detection code and eliminating DummyAnalyzer added earlier .
> > > 
> > > 
> > > great thanks, were you able to run it on windows?
> > 
> > 
> > yes
> 
> perfect, so apart from the steps/commands I suggested, did you have to do anything else?

no 
the command provided was enough .

aatmanvaidya commented:
hi @hooiv great work with the PR, thank you for your patience in making the constant edits. 
I have made some minor fixes, there were some issues in the test, merging the PR now
